### PR TITLE
[Preview 6] Use JsonReader interop with JsonSerializer

### PR DESCRIPTION
**Pull request by BrennanConroy**
- Fix up some tests

Before:

Method |          Input |    HubProtocol |        Mean |        Error |       StdDev |        Op/s |  Gen 0 |  Gen 1 | Allocated |
------------------ |--------------- |--------------- |------------:|-------------:|-------------:|------------:|-------:|-------:|----------:|
 Read |   FewArguments |           Json |  3,157.1 ns |    65.910 ns |   134.638 ns |   316,749.1 | 0.0153 |      - |     480 B |
 Read | LargeArguments |           Json | 44,724.6 ns | 2,950.102 ns | 8,605.586 ns |    22,359.1 | 4.2114 | 0.1221 |  102800 B |
 Read |  ManyArguments |           Json |  6,184.8 ns |   123.451 ns |   121.245 ns |   161,687.4 | 0.0305 |      - |    1056 B |
 Read |    NoArguments |           Json |  1,441.7 ns |    28.604 ns |    25.356 ns |   693,640.3 | 0.0057 |      - |     176 B |

After:

Method |          Input |    HubProtocol |        Mean |        Error |       StdDev |        Op/s |  Gen 0 |  Gen 1 | Allocated |
------------------ |--------------- |--------------- |------------:|-------------:|-------------:|------------:|-------:|-------:|----------:|
 Read |   FewArguments |           Json |  2,147.6 ns |    42.544 ns |   114.293 ns |   465,638.8 | 0.0038 |      - |     224 B |
 Read | LargeArguments |           Json | 16,928.1 ns |   706.900 ns | 2,039.569 ns |    59,073.3 | 1.0834 | 0.0153 |   30904 B |
 Read |  ManyArguments |           Json |  4,251.8 ns |    83.215 ns |   102.196 ns |   235,193.3 | 0.0153 |      - |     424 B |
 Read |    NoArguments |           Json |    750.6 ns |    14.886 ns |    30.071 ns | 1,332,260.4 | 0.0029 |      - |      96 B |

Newtonsoft and MsgPack for comparison:

Method |          Input |    HubProtocol |        Mean |        Error |       StdDev |        Op/s |  Gen 0 |  Gen 1 | Allocated |
------------------ |--------------- |--------------- |------------:|-------------:|-------------:|------------:|-------:|-------:|----------:|
 Read |   FewArguments |        MsgPack |    629.2 ns |    12.395 ns |    13.777 ns | 1,589,257.0 | 0.0076 |      - |     224 B |
 Read |   FewArguments | NewtonsoftJson |  2,281.4 ns |    44.465 ns |    76.699 ns |   438,329.8 | 0.0267 |      - |     960 B |
 Read | LargeArguments |        MsgPack |  7,936.3 ns |   498.136 ns | 1,460.948 ns |   126,003.9 | 1.0910 | 0.0229 |   30904 B |
 Read | LargeArguments | NewtonsoftJson | 68,549.2 ns | 1,352.304 ns | 3,291.694 ns |    14,588.1 | 1.9531 |      - |   58840 B |
 Read |  ManyArguments |        MsgPack |  1,099.8 ns |    22.493 ns |    18.783 ns |   909,275.0 | 0.0114 |      - |     400 B |
 Read |  ManyArguments | NewtonsoftJson |  4,164.0 ns |    82.231 ns |   123.080 ns |   240,151.6 | 0.0458 |      - |    1504 B |
 Read |    NoArguments |        MsgPack |    318.3 ns |     6.272 ns |     7.222 ns | 3,141,842.9 | 0.0038 |      - |     120 B |
 Read |    NoArguments | NewtonsoftJson |  1,016.5 ns |    26.047 ns |    32.941 ns |   983,747.3 | 0.0191 |      - |     584 B |

cc @ahsonkhan

### CODE REVIEW

File: src/SignalR/clients/ts/FunctionalTests/ComplexObject.cs
Comment (davidfowl) [inline src/SignalR/clients/ts/FunctionalTests/ComplexObject.cs]: GUID -> Guid

File: src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts
Comment (BrennanConroy) [inline src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts]: These test changes aren't random. Re-adding coverage now that features exist: https://github.com/aspnet/AspNetCore/commit/c84e37f30def9ff0b2d12e877e3de6c283be6145#diff-1c30bdd79e21ac4ce12ab9663aa438e0L508

**Comment by davidfowl** [comment]
Very nice! MsgPack still crushing JSON though zomg.

File: src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs
Comment (rynowak) [inline src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs]: This is kinda unfortunate. @ahsonkhan - is this the best pattern for "I maybe have a reader"?

File: src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs
Comment (rynowak) [inline src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs]: L702 can be simplified and brought inside the look for more pranav points. Then you avoid the allocation with a zero-parameter method in the case that there's data in the array.

File: src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs
Comment (rynowak) [inline src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs]: Do we need to test examples with invalid nesting now that this use the reader?

**Comment by BrennanConroy** [comment]
🆙 📅

**Comment by BrennanConroy** [comment]
Ping @halter73 @mikaelm12 we might get QB approval for this so review/approval would be nice (read necessary)

File: src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs
Comment (ahsonkhan) [reply src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs]: Yes, given its a struct. We also do not have the equality operators/IsDefault API since there wasn't a motivating scenario which showcased that such an API would add value: https://github.com/dotnet/corefx/issues/37534

**Comment by anurse** [comment]
Please do review ASAP. Submitted for QB Approval.

**Comment by anurse** [comment]
Approved by QB for preview 6. Please get review approval and passing build before merging.

File: src/SignalR/clients/csharp/Client/test/FunctionalTests/HubConnectionTests.cs
Comment (mikaelm12) [inline src/SignalR/clients/csharp/Client/test/FunctionalTests/HubConnectionTests.cs]: This probably only needs to be varied over the different HubProtocols but meh

File: src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs
Comment (BrennanConroy) [reply src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs]: Yes https://github.com/aspnet/AspNetCore/blob/74d9031f12c4ed1535f35caa92683d57e599ddfb/src/SignalR/common/SignalR.Common/test/Internal/Protocol/JsonHubProtocolTestsBase.cs#L173

File: src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs
Comment (BrennanConroy) [reply src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs]: Maybe, but this is unrelated to the changes. Let's try not to have extra unneeded churn. File an issue or open a PR 😄

File: src/SignalR/clients/csharp/Client/test/FunctionalTests/HubConnectionTests.cs
Comment (halter73) [reply src/SignalR/clients/csharp/Client/test/FunctionalTests/HubConnectionTests.cs]: Good call.

File: src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts
Comment (halter73) [inline src/SignalR/clients/ts/FunctionalTests/ts/HubConnectionTests.ts]: Is this recorded as a breaking change?

File: src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs
Comment (halter73) [reply src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs]: Is it too late to consider this a motivating scenario?

**Review by halter73 — APPROVED** [review]

**Comment by anurse** [comment]
@BrennanConroy are we blocked on anything other than dependencies flowing in now? Looks like no (builds are red because of that and we have review approval). Just want to confirm where we stand.

**Comment by BrennanConroy** [comment]
> are we blocked on anything other than dependencies flowing in now? Looks like no

That's correct, I might change the test commented on above, but that isn't blocking

**Comment by anurse** [comment]
My feeling is that we should just cherry-pick this over to the dependency update PR when it arrives, but I'm not sure. The other option is to receive the dependencies and then rebase/update this PR when they arrive. That would mean another PR validation turn before we can merge it though. (Thoughts @aspnet/build)

**Comment by dougbu** [comment]
@anurse my leaning is strongly toward pushing a commit or two on top of what the dependency bot comes up with. Just need to wait for aspnet/EntityFrameworkCore#15940 and the official build including that.

BTW why does this PR say it's about "Preview 6" but target 'master'?

**Comment by BrennanConroy** [comment]
I was always planning on cherry-picking

**Comment by anurse** [comment]
Spoke with @natemcmaster . Let's do this the "right" way and cherry-pick only the change needed to unblock the dependency PR there (there *is* a breaking change that we need to react to).

Then, we rebase this PR on top of the newly-arrived corefx dependencies (when they hit the release branch), retarget the PR and get our PR validation build green.

**Comment by natemcmaster** [comment]
Dependency flow should be available soon. https://github.com/aspnet/AspNetCore-Tooling/pull/650 just merged.

Once [this build is done](https://dev.azure.com/dnceng/internal/_build/results?buildId=211838) and published to Maestro, I would recommend updating your PR to use latest deps by running `darc update-dependencies --channel '.NET Core 3 Release'` and pushing the changes to Version.props/xml to your PR branch. The bot is still going to open a separate PR to flow dependency versions, but this will help you get PR checks running with updated dependencies before we complete the bot dependency flow into AspNetCore (and spot any problems that might exist)

**Comment by dougbu** [comment]
Note
> Dependency flow should be available soon. aspnet/AspNetCore-Tooling#650 just merged.

That's the wrong repo to watch for important updates from Extensions or lower. We're flowing dependencies through EF Core again both here and in 'master'.

In any case, I believe aspnet/AspNetCore#10856 is the dependency update PR of interest.

**Review by mikaelm12 — APPROVED** [review]

**Comment by BrennanConroy** [comment]
@aspnet/build Someone merge please :)
Helix is hanging it looks like.

**Comment by anurse** [comment]
[QB approved](https://github.com/aspnet/AspNetCore/pull/10733#issuecomment-498472202) and good to go. Merging away!

File: src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs
Comment (ahsonkhan) [reply src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs]: > Is it too late to consider this a motivating scenario?

If we are asking for an `IsDefault`, not really, especially since this is quite a trivial API change. The only concern is that the impact of such an API is quite low, so I consider this really low priority (we can always add it in the future). Unless creating multiple `Utf8JsonReader` structs (as default), up front, is affecting performance and having an `IsDefault` check would help, I wouldn't prioritize fixing it for 3.0. In this case (we have several readers), so it might actually affect perf noticeably. 

@BrennanConroy, do you have a way to test the perf delta assuming such a property exists?

Also, wouldn't the compiler complain if you did this:
```C#
Utf8JsonReader resultToken;
if (resultToken.IsDefault) // Using unassigned variable.
{
 // ...
}
```

So, the question is what does the use case/scenario look like here? Do you have a proposal in mind? Looking more closely, it doesn't look like `IsDefault` would work for this case anyway?

**Pull request merged**
